[ADD][11.0] fieldservice_recurring

brian10048 opened the pull request:
A start to #164 
Open to suggestions/modifications.  Hopefully this can meet all of our needs.

Still needed:
- [x] use a maximum number of occurrences to limit new order creation
- [x] add days of week schedule preferences
- [x] manage state for renew and expiration of fsm.recurring thru its cron task

brian10048 committed:
[ADD] fieldservice_recurring

wolfhall commented:
Brian,
this is a very nice addition - a few comments:
Currently, the frequencies are hard coded for week, month, etc.  
Some customers require semi annual or quarterly, or twice a week, etc.

This may be o.k. for now but in the future, we should have a frequency configuration that defines:

- Frequency name: the name for this frequency, e.g. Weekly - exact, or Monthly +/- 4 days
- Frequency default interval, Integer: the default number of days between events. e.g. 7 for weekly
- Frequency earliest buffer: the number of days to arrive prior to the exact calculated date. 
- Frequency latest buffer: the number of days to arrive after to the exact calculated date. 
- The last two values define the earliest and latest start date for a job while the default interval defines the scheduled date. this was the dispatcher has the flexibility to move these orders around.

This way the user can define their own frequencies and the module will calculate the earliest and latest start dates.

We also may want to add a configuration parameter: 
Schedule horizon: defines the maximum number of days that should be scheduled out in advance.
A scheduled action will look at the value and create new orders when needed.

brian10048 commented:
I see how this falls short for more configurable frequencies and I can work on that.

The configuration for a schedule horizon is in here.  It defaults to 30 `days_ahead` in the model's cron method parameters.  If needed, the user can set this parameter value via the configuration of the scheduled action.

brian10048 committed:
[IMP] fieldservice_recurring

FIX travis for format
ADD frequency model for dates

brian10048 commented:
Still work to be done here, but this should add the configurable Frequency as requested 🥂

brian10048 commented on fieldservice_recurring/models/fsm_recurring.py:
```diff
@@ -0,0 +1,198 @@
+            'location_id': self.location_id.id,
+            'scheduled_date_start': schedule_date,
+            'request_early': str(earliest_date),
+            'request_late': str(latest_date),
```
This isn't setting the value on the request late field as expected because it is a computed field of fsm.order

babatoko commented:
Hello, great start indeed and as you mention it's a work in progress.

I didn't figure out yet how I can set worker order frequency in my following use case.

One of my customer request me to service him every of weekdays except for even week we work 4 days (M,T,T,F).

Would be great to implement some sort of exclusion mechanism.

In my business, we rely strongly on time based on customers business hours.

As Example, a work have to be done after business hours and for each customers the closing time may be different. A constraint should be set for when creating work order template or work order a check occur on start and end time.

brian10048 committed:
[IMP] recurring frequency

Added ability to calculate recurring dates using rrule from dateutil python library

brian10048 commented:
This now uses the dateutil rrule python library to calculate order dates.  Readme updated to help understand its usage. 

Please review and recommend changes!

osi-scampbell commented:
Hey @brian10048  the code looks really good. A few things to help pass the Travis-ci build. On fsm_recurrying.py lines 48 and 49 need a space before them so they line up. It looks like fsm_recurring_template.py lines 18 and 19 need two spaces. That should fix the Flake8 errors in the build.

brian10048 committed:
[IMP] fieldservice recurring

Fixed Travis errors
Added early, late buffer for order requests
Improved recurring order cron expiration method
Minor UI improvements

brian10048 commented:
@osi-scampbell Travis errors are fixed.
Also some other minor improvements.  I believe I have done everything initially planned for this addition.  Let me know if any other modifications are needed.  If not, I can squash these into one commit before merging.

brian10048 committed:
[FIX] recurring readme

update for Travis errors

babatoko commented:
I run a test of the new commit and here what I found out.

```
Error:
Odoo Server Error

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/odoo/tools/safe_eval.py", line 350, in safe_eval
    return unsafe_eval(c, globals_dict, locals_dict)
  File "", line 1, in <module>
  File "/var/local/odoo/field-service/fieldservice_recurring/models/fsm_recurring.py", line 235, in _cron_scheduled_task
    self._cron_generate_orders()
  File "/var/local/odoo/field-service/fieldservice_recurring/models/fsm_recurring.py", line 182, in _cron_generate_orders
    lambda r: datetime.strptime(r.scheduled_date_start,
  File "/usr/lib/python3/dist-packages/odoo/models.py", line 4542, in mapped
    return self._mapped_func(func)
  File "/usr/lib/python3/dist-packages/odoo/models.py", line 4523, in _mapped_func
    vals = func(self)
  File "/var/local/odoo/field-service/fieldservice_recurring/models/fsm_recurring.py", line 183, in <lambda>
    DEFAULT_SERVER_DATETIME_FORMAT
TypeError: strptime() argument 1 must be str, not bool

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 651, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 310, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/usr/lib/python3/dist-packages/odoo/tools/pycompat.py", line 87, in reraise
    raise value
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 693, in dispatch
    result = self._call_function(**self.params)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 342, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/service/model.py", line 97, in wrapper
    return f(dbname, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 335, in checked_call
    result = self.endpoint(*a, **kw)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 937, in __call__
    return self.method(*args, **kw)
  File "/usr/lib/python3/dist-packages/odoo/http.py", line 515, in response_wrap
    response = f(*args, **kw)
  File "/usr/lib/python3/dist-packages/odoo/addons/web/controllers/main.py", line 938, in call_button
    action = self._call_kw(model, method, args, {})
  File "/usr/lib/python3/dist-packages/odoo/addons/web/controllers/main.py", line 926, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/usr/lib/python3/dist-packages/odoo/api.py", line 689, in call_kw
    return call_kw_multi(method, model, args, kwargs)
  File "/usr/lib/python3/dist-packages/odoo/api.py", line 680, in call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/odoo/addons/base/ir/ir_cron.py", line 76, in method_direct_trigger
    self.sudo(user=cron.user_id.id).ir_actions_server_id.run()
  File "/usr/lib/python3/dist-packages/odoo/addons/base/ir/ir_actions.py", line 554, in run
    res = func(action, eval_context=eval_context)
  File "/usr/lib/python3/dist-packages/odoo/addons/base/ir/ir_actions.py", line 430, in run_action_code_multi
    safe_eval(action.sudo().code.strip(), eval_context, mode="exec", nocopy=True)  # nocopy allows to return 'action'
  File "/usr/lib/python3/dist-packages/odoo/tools/safe_eval.py", line 373, in safe_eval
    pycompat.reraise(ValueError, ValueError('%s: "%s" while evaluating\n%r' % (ustr(type(e)), ustr(e), expr)), exc_info[2])
  File "/usr/lib/python3/dist-packages/odoo/tools/pycompat.py", line 86, in reraise
    raise value.with_traceback(tb)
  File "/usr/lib/python3/dist-packages/odoo/tools/safe_eval.py", line 350, in safe_eval
    return unsafe_eval(c, globals_dict, locals_dict)
  File "", line 1, in <module>
  File "/var/local/odoo/field-service/fieldservice_recurring/models/fsm_recurring.py", line 235, in _cron_scheduled_task
    self._cron_generate_orders()
  File "/var/local/odoo/field-service/fieldservice_recurring/models/fsm_recurring.py", line 182, in _cron_generate_orders
    lambda r: datetime.strptime(r.scheduled_date_start,
  File "/usr/lib/python3/dist-packages/odoo/models.py", line 4542, in mapped
    return self._mapped_func(func)
  File "/usr/lib/python3/dist-packages/odoo/models.py", line 4523, in _mapped_func
    vals = func(self)
  File "/var/local/odoo/field-service/fieldservice_recurring/models/fsm_recurring.py", line 183, in <lambda>
    DEFAULT_SERVER_DATETIME_FORMAT
ValueError: <class 'TypeError'>: "strptime() argument 1 must be str, not bool" while evaluating
'model._cron_scheduled_task()'
```

When I set a new frequency, there is a constraint error on Day of month field witch seems to be set to 0 even when the field is invisible.

wait for a fix to go further with testing.

brian10048 committed:
[FIX] recurring

improve cron method to prevent errors
improve constrains method to prevent errors

brian10048 commented:
@babatoko I have tested again and I am having success using the cron task.

Did you 'Confirm' the Recurring Order? Cron task will not create orders for Recurring Orders in the draft state.
It may also be that your Frequency Rule Set is configured in a way that no dates are available to create the orders.

I tested this on the current runbot build [link here](http://3366643-179-476bcf.runbot1.odoo-community.org/) Please login and check it out
I created your use case of daily on weekdays, excluding the Wednesdays of even number weeks.  Please see the Frequency Rule Set configuration I made for this. 
You can then look at the Recurring Order "RFO002" which shows the expected result after Confirming the Recurring Order, and then running Cron task.  

If I need to improve the readme to provide better instructions or you are experiencing other issues please let me know!

babatoko commented:
I don't have access to runbot. How I can get one ?

brian10048 commented:
Use admin as username and password at the link in previous comment

brian10048 committed:
Merge branch '11.0' into dev-rec

max3903 commented on fieldservice_recurring/models/fsm_frequency.py:
```diff
@@ -0,0 +1,150 @@
+    interval = fields.Integer(
+        string='Repeat Every', help="The number of intervals between events",
+        default=1, required=True, track_visibility='onchange')
+    interval_type = fields.Selection([
```
```suggestion
    interval_type = fields.Selection(FREQUENCIES,
```

Can you use the variable defined above?
Like here:
https://github.com/OCA/field-service/blob/12.0/fieldservice_stock/models/fsm_order.py#L43

brian10048 replied:
Done [https://github.com/OCA/field-service/pull/179/commits/709bcbd9507ea0eca154731a3babd78a89918429](https://github.com/OCA/field-service/pull/179/commits/709bcbd9507ea0eca154731a3babd78a89918429)

max3903 commented:
@brian10048 

Can you add an option in the configuration panel to install this module? 

The module should also add the template group as an inherited group of the Recurring one, don't you think? If you want to manage recurring orders, you want to manage templates also, I think...

max3903 reviewed (changes requested):
Minor changes and question

brian10048 committed:
[FIX] recurring

use variable defined for selection field

brian10048 commented:
@max3903 

> 
> Can you add an option in the configuration panel to install this module?
> 

In res_config_settings?  It is already there

>
> The module should also add the template group as an inherited group of the Recurring one, don't you think? If you want to manage recurring orders, you want to manage templates also, I think...
>

I originally thought one would want to manage the templates as well, but decided to make it functional for either with or without the template group enabled. I don't mind setting it up to enable the template group if that's what we should do

babatoko commented:
May be a bit late, but as all things seems to move on 12 version, would it be to port up this module to 12 as well?

Thanks

max3903 commented:
@babatoko Possible? Yes. The module is in #15.
